PSPs/drafts/psp-34.md

```diff
@@ -0,0 +1,914 @@
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "label": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Transfer approved or owned token from `from`.",
+    "",
+    "On success a `Transfer` event is emitted.",
+    "",
+    "# Errors",
+    "",
+    "Returns `TokenNotExists` error if `id` does not exist.",
+    "",
+    "Returns `NotApproved` error if `from` doesn't have allowance for transferring.",
+    "",
+    "Returns `SafeTransferCheckFailed` error if `to` doesn't accept transfer."
+  ],
+  "mutates": true,
+  "label": "PSP34::transfer_from",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 1
+  },
+  "selector": "0x718fd38b"
+}
+```
+
+#### PSP34Receiver
+`PSP34Receiver` is an interface for any contract that wants to support safe transfers from a PSP-34 token smart contract to avoid unexpected tokens in the balance of contract.
+This method is called before a transfer to ensure the recipient of the tokens acknowledges the receipt.
+
+##### **before_received**(operator: AccountId, from: AccountId, id: Id, data: [u8]) ➔ Result<(), PSP34ReceiverError>
+Selector: `0xbb7df780` - first 4 bytes of `blake2b_256("PSP34Receiver::before_received")`
+```json
+{
+  "args": [
+    {
+      "label": "operator",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "label": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Ensures that the smart contract allows reception of PSP34 token(s).",
+    "Returns `Ok(())` if the contract allows the reception of the token(s) and Error `TransferRejected(String)` otherwise.",
+    "",
+    "This method will get called on every transfer to check whether the recipient in `transfer`",
+    "or `transfer_from` is a contract, and if it is, does it accept tokens.",
+    "This is done to prevent contracts from locking tokens forever.",
+    "",
+    "Returns `PSP34ReceiverError` if the contract does not accept the tokens."
+  ],
+  "mutates": true,
+  "label": "PSP34Receiver::before_received",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 2
+  },
+  "selector": "0xbb7df780"
+}
+```
+
+### Extension
+
+#### PSP34Metadata
```

Comment:
The proposed Metadata extension differs quite a lot from ERC721.
Why is there no `tokenURI` function?

Reply:
Because attributes are more flexible than a single function. The contract provides only one method to get attributes and anyone can get any attribute. For example, `get_atribute(Id::U8(1), "token_uri")` can return `tokenURI` for some NFT. But it also can provide some other attributes like `image`, `type` and etc.

Reply:
how are attributes set?

Reply:
The setting of attributes is permissioned operation. So the creator of the contract should define the way how he will set attributes. For example, during minting, he also can pass the list of attributes.

Reply:
Ok, that makes sense.
Maybe it would be interesting to add a few words into the metadata description explaining the role of attributes?

That way, people that are coming from ERC721 familiarity can easily understand how attributes are a more flexible way to do the same things they're used to.